server/src/omtLinkProvider.ts

```diff
@@ -0,0 +1,174 @@
+import { dirname, isAbsolute, resolve } from 'path';
+import { DocumentLink, Position, Range } from 'vscode-languageserver';
+import { TextDocument } from 'vscode-languageserver-textdocument';
+import { readFileSync } from 'fs';
+import { glob } from 'glob';
+import { WorkspaceLookup } from './workspaceLookup';
+import { DeclaredImportLinkData, isDeclaredImportLinkData } from './types';
+
+const omtUriMatch = /( +["']?)(.*\.omt)/;
+const declaredImportMatch = /( +)(module:)(.*):/;
+const importMatch = /^import:/g;
+const otherDeclareMatch = /^(\w+):/g;
+
+/**
+ * Provides DocumentLinks for the imports of an OMT file.
+ */
+export default class OMTLinkProvider {
+
+    constructor(private workspaceLookup: WorkspaceLookup) { }
+
+    provideDocumentLinks(document: TextDocument): DocumentLink[] {
+        // regular path links, with or without shorthands
+        const shorthands = this.contextPaths(document);
+        return findOMTUrl(document, (uri) => replaceStart(uri, shorthands));
+    }
+
+    /**
+     * Resolve the target of a DocumentLink using its data. Returns undefined when unable to resolve
+     * @param data the data from a document link
+     */
+    resolveLink(data?: DeclaredImportLinkData | unknown) {
+        if (data && isDeclaredImportLinkData(data)) {
+            return this.workspaceLookup.getModulePath(data.declaredImport.module);
+        }
+    }
+
+    contextPaths(document: TextDocument) {
+        const pathPaths = glob.sync('**/tsconfig**.json')
+            // we need to filter for same parent because another config may have some of the same paths
+            .filter(uri => document.uri.startsWith(uri.substr(0, uri.lastIndexOf('/'))))
+            .map(uri => {
+                try {
+                    const file = readFileSync(uri, 'utf8');
+                    const json = JSON.parse(file);
+                    if (json.compilerOptions && json.compilerOptions.paths) {
+                        // now make a path from the config folder to the keys value and map that
+                        const paths: [string, string][] = [];
+                        for (const key in json.compilerOptions.paths) {
+                            const relPath = json.compilerOptions.paths[key].toString();
+                            const newPath = resolve(dirname(uri), relPath);
+                            paths.push([key.substr(0, key.lastIndexOf('/*')), newPath]);
+                        }
+                        return paths;
+                    }
```

Comment:
here needs a return when the condition is else

Reply:
it was an implicit undefined return, I'll make it explicit. The filter in the end already takes care of those values.

Reply:
I will be using the reduce approach suggested by emiel later in this threat